.travis.yml

```diff
@@ -0,0 +1,56 @@
+stages:
+  - unit_test
+  - integration_test
+
+language: java
+
+services:
+  - docker
+
+jdk:
+  - openjdk11
+
+os:
+  - linux
+
+cache:
+  directories:
+    - ~/.m2/repository
+
+install:
+  - mvn install -DskipTests=true -Dmaven.javadoc.skip=true --quiet --batch-mode --show-version --activate-profiles travis
+
+.base_unit_test: &base_unit_test
+  stage: unit_test
+  if: type = push
+  script:
+    - mvn -pl "$TEST_SUITE" test
+
+.base_integration_test: &base_integration_test
+  stage: integration_test
+  if: type = pull_request
+  script:
+    - mvn -pl 'beekeeper-integration-tests' -Dtest=$TEST_FILE test
+
+jobs:
```

Comment:
I think you could do all of the above using surefire and/or maven include/exclude rules for *IntegrationTest. Having said that, do we need this separation of unit and integration tests? Shouldn't we have the CI server run *everything* every time - package, coverage, checkstyle, findbugs and all the different tests? It's less to configure here and we don't mind if this takes a bit longer to run in the background (it can take a while to acquire and spin up VMs anyway).

Reply:
Answering this and the comment for `if: type = push`, `if: type = pull_request` here.

Essentially the way I have this pipeline setup is to run the full suite of unit tests on push, but only run integration tests on open PRs.  I've got all the unit test suites broken up as well to run them in separate processes in Travis so they'll report failures individually. Overall It's more of a choice to keep build times for commits quicker so if you're developing a feature you'll have a faster feedback loop to see that the change you made has broken something before you open up a PR. 

Generally I've seen in practice teams do integration testing on only PR events or after merging to master (as part of an automated deployment setup) since integration tests usually are a large factor of time slower than unit tests.

Although with all that said - it really comes down to preference of what workflow works best for our team, so if we want everything to just run on every commit that is totally fine too 😄

Reply:
For dev feedback I recommend running the unit tests locally. I think CI should run everything and be more of a safety net when you're asking others to review your code, which is usually a bunch of commits rolled up into a push. I definitely have a preference to just run the whole Maven build each time this happens as that's what we've been doing up to now in other projects and it seems to work well and everyone is happy with that. Speak up if anyone else disagrees ;)

Reply:
Reverting that now to always run the full test suite on commit 🚀 

It looks like I'm running into some issues w/ the integration suite now though where it's not completing some tests within the allotted time which is a bit strange. Looking into that some more now

Reply:
Let's see what Travis thinks of this, we can always update this later...